ImmuDB4Net/ImmuClientSync.cs

```diff
@@ -0,0 +1,1364 @@
+    }
+
+    public class LibraryWideSettings
+    {
+        public int MaxConnectionsPerServer { get; set; }
+        public TimeSpan TerminateIdleConnectionTimeout { get; set; }
+        public TimeSpan IdleConnectionCheckInterval { get; set; }
+        internal LibraryWideSettings()
+        {
+            //these are the default values
+            MaxConnectionsPerServer = 2;
+            TerminateIdleConnectionTimeout = TimeSpan.FromSeconds(60);
+            IdleConnectionCheckInterval = TimeSpan.FromSeconds(6);
+        }
+    }
+
+    public ImmuClientSync() : this(NewBuilder())
+    {
+
+    }
+
+    public ImmuClientSync(string serverUrl, int serverPort)
+        : this(NewBuilder().WithServerUrl(serverUrl).WithServerPort(serverPort))
+    {
+    }
+
+    public ImmuClientSync(string serverUrl, int serverPort, string database)
+        : this(NewBuilder().WithServerUrl(serverUrl).WithServerPort(serverPort).WithDatabase(database))
+    {
+    }
+
+    internal ImmuClientSync(ImmuClientSyncBuilder builder)
+    {
+        ConnectionPool = builder.ConnectionPool;
+        GrpcAddress = builder.GrpcAddress;
+        connection = releasedConnection;
+        SessionManager = builder.SessionManager;
+        DeploymentInfoCheck = builder.DeploymentInfoCheck;
+        serverSigningKey = builder.ServerSigningKey;
+        stateHolder = builder.StateHolder;
+        heartbeatInterval = builder.HeartbeatInterval;
+        ConnectionShutdownTimeout = builder.ConnectionShutdownTimeout;
+        stateHolder.DeploymentInfoCheck = builder.DeploymentInfoCheck;
+        stateHolder.DeploymentKey = Utils.GenerateShortHash(GrpcAddress);
+        stateHolder.DeploymentLabel = GrpcAddress;
+    }
+
+
+    public static void ReleaseSdkResources()
+    {
+        RandomAssignConnectionPool.Instance.Shutdown();
+    }
+
+    private void StartHeartbeat()
+    {
+        heartbeatTask = Task.Factory.StartNew(async () =>
+        {
+            while (true)
+            {
+                if (heartbeatCloseRequested!.WaitOne(heartbeatInterval)) return;
+                try
+                {
+                    await Service.KeepAliveAsync(new Empty(), Service.GetHeaders(ActiveSession));
+                    heartbeatCalled?.Set();
+                }
+                catch (RpcException) { }
+            }
+        });
+    }
+
+    private void StopHeartbeat()
+    {
+        if (heartbeatTask == null)
+        {
+            return;
+        }
+        heartbeatCloseRequested?.Set();
+        heartbeatTask.Wait();
+        heartbeatCloseRequested?.Close();
+        heartbeatCalled?.Close();
+        heartbeatTask = null;
+    }
+
+    public void Open(string username, string password, string defaultdb)
+    {
+        try
+        {
+            using (ManualResetEvent mre = new ManualResetEvent(false))
+            {
+                while (true)
+                {
+                    if (Interlocked.Exchange(ref sessionSetupInProgress, 1) == 0)
```

Comment:
I think we don't need this trick in sync code anymore, correct? We can use just a single lock here instead that IMO would greatly simplify the code.